doc/notes/188_enhance_logging_ant.adoc

```diff
@@ -0,0 +1,369 @@
+messages.  The most common destination for a log message is a file.
+Another type of appender directs log entries to `syslogd`.  There is
+support for appending to a Kafka topic.  This is useful in the Protocol
+Verifier.  For example, Logger entries bound for Reporting can be assigned
+to topics.  Likewise, benchmarking can be enabled, disabled and destined
+in this way.  A logger can configured with multiple appenders.  Appenders
+have functionality to filter what gets logged based upon log level,
+textual strings and more.
+
+=== 2.2 Logging Best Practices
+
+As preparation for enhancing logging, a few published "Best Practice"
+guides were studied.  We will take the opportunity in Munin to adopt/apply
+some of these practices while changing to the `log4*` idiom of logging.
+
+==== 2.2.1 Mission Statement
+
+The first Best Practice to logging is to author a mission statement for
+the logging and proceed from there.  This aligns with Shlaer-Mellor
+domain modelling quite nicely.
+
+==== 2.2.2 Canonical Logging
+
+Canonical logging is the practice of strategically delivering log lines in
+a manner that eases the analysis of the logged data.  The technique
+accumulates multiple related key-value pairs and logs them together in a
+single entry.  An example canonical log message in the Protocol Verifier
+will be the 'job failed' log message.  This will be an assembly of all of
+the information needed to understand the failure including a trace of the
+events composing the failed job.
+
+== 3 Requirements
+
+. Establish a Project Munin logging mission statement to guide decision
+  making.
+. Design Reporting to flow from Logging.
+  ** Reporting shall be mechanically identical to Logging.
+  ** Reporting shall be a subset of Logging.
+  ** Reporting shall be a configured set of loggers and log levels.
+. Support logging levels:
+  ** Fatal
+  ** Error
+  ** Warning
+  ** Information
+  ** Debug
+  ** Trace
+. Support conditional log message string arithmetic.
+. Establish conventions for enabled logging.
+. Define and support a logger hierarchy.
+. Establish naming conventions for the logger hierarchy.
+. Establish rules and guidelines for the construction of log messages.
+. Facilitate machine parsing of log messages (using structured logging).
+  This will result in logs that can be easily ingested by log analysis
+  tooling.
+. Identify tooling that will be used to harvest the log data.
+. Protect privacy and security by logging application-level information
+  judiciously.
+
+== 4 Analysis
+
+=== 4.1 Munin Logging Mission Statement
+
+* Munin software shall log information critical to the verification of protocols.
+  This includes detection of alarms, failures and successes.
+* Munin software shall be instrumented to enable the measurement of the
+  performance of the Protocol Verifier application itself (benchmarking).
+* Munin software shall be instrumented sparingly with debug statements
```

Comment:
Is this important if (a) there are different logging levels and (b) the ability for 'lower' levels of logging to be switched off? - for example, if everything at or lower than 'debug' is switched off then those log messages will not 'flood' the PV logging but if needed for detailed inspection they can be switched back on again.